ast.go

```diff
@@ -3770,42 +3771,78 @@ func (e *Wildcard) String() string {
 
 // CloneExpr returns a deep copy of the expression.
 func CloneExpr(expr Expr) Expr {
+	// Call CloneExprWithConstraints but with basically no constraints -- a
+	// background context which will carry no cancellation information and 0
+	// maximum depth which means infinite recursion.
+	return func() Expr { e, _ := CloneExprWithConstraints(context.Background(), 0, expr); return e }()
+}
+
+func CloneExprWithConstraints(ctx context.Context, maxDepth int, expr Expr) (Expr, error) {
```

Comment:
Instead of making a function for the constraints specifically, can you do this pattern?

```
type CloneOption interface {
    apply(opts *cloneOptions)
}

type cloneOptions struct {
    maxDepth int
}

type cloneOptionFn func(opts *cloneOptions)

func (fn cloneOptionFn) apply(opts *cloneOptions) {
    fn(opts)
}

func WithMaxDepth(n int) CloneOption {
    return cloneOptionFn(func(opts *cloneOptions) {
        opts.maxDepth = n
    })
}

func CloneExprWithContext(ctx context.Context, expr Expr, opts ...CloneOption) (Expr, error) {
}

I like this pattern more because I find it to be more extensible.

Reply:
So --

Typically I see this done in "factory" type functions when we try to hide implementation details of whatever type we're building.  In that case it is reasonable to account for unforeseen changes.

In this case I feel like we have a pretty clear view of what we want to do:  We want to respect cancellation signals via the `context.Context` and we want to limit the recursion depth.

I can't really foresee anything else I'd like to do with `CloneExpr()`.  I imagine there might be a change required in the future but, in this case, it'd be my inclination to keep it simple for now.

Clearly it is nice to make room for future enhancements this way but in this case I'm willing to compromise flexibility for the sake for simplicity -- it just doesn't feel like the right time or place to make an arbitrarily extensible routine.

Reply:
Another reason why I like it is because it makes it more self-documenting. It's like adding a keyword argument. I'm not a big fan of the `CloneExprWithConstraints` because they aren't really multiple constraints. It's one constraint for max depth and it doesn't really say that. The primary difference between this function and `CloneExpr` is to pass the context and the typical way to do this is to use `WithContext` as the suffix.

I guess if you didn't want to add the option syntax, we could go with `CloneExprWithContext` and just have `maxDepth` as an option. I don't like the constraints name though.

Reply:
Okay -- that's fine.  There are multiple constraints in `CloneExprWithConstraints()`... the context and the depth.

I get it -- I think I like the functional options now.